### PR TITLE
Cache with prefetch

### PR DESCRIPTION
**Pull request by vidhya567**
-A separate angular cache module is created
-In mapcityJson function just calls the getCityPolygonsFromCache function of cache,every other function of Cache is abstracted away from map
**Cache Prefetch**
Based on trend angle
**Cache replacement**
-use the trend angle to find direction of eviction
-divide the cachedRegion into parts excluding the overlapping part of current Polygon region request
-cut the bigger parts first see whether deleteTarget is achieved
-if not cut the next part

### CODE REVIEW

File: twittermap/public/javascripts/app.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
-var app = angular.module('cloudberry', ['cloudberry.map','cloudberry.timeseries','cloudberry.sidebar', 'cloudberry.util']);
+var app = angular.module('cloudberry', ['cloudberry.map','cloudberry.timeseries','cloudberry.sidebar', 'cloudberry.util','cloudberry.cache']);
 
 app.controller("AppCtrl", function ($scope) {
-  
+
```
Comment (waans11) [inline twittermap/public/javascripts/app.js]: This blank line is not necessary.

File: twittermap/public/javascripts/cache/services.js [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+/*Cache module stores user requested city polygons.When Users requests these same city polygon area again we have it in cache
+and provide it to the user without sending http request.*/
+angular.module('cloudberry.cache', ['leaflet-directive', 'cloudberry.common' ])
+.service('Cache', function( $window, $http, $compile){
+
+var cachedCityPolygonTree = rbush();
+var cachedRegion ;
+var cacheSize = 0;
+var insertedTreeIDs = new Set();
+var cacheThreshold = 7500;//hard limit
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Did you measure the MB size for this?  I wonder whether this is a reasonable size.

File: twittermap/public/javascripts/cache/services.js [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+var targetDeleteCount = 0;
+var deletedPolygonCount = 0;
+var preFetchDistance = 25;
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Let's put a comment: // mile

File: twittermap/public/javascripts/cache/services.js [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+
+function getAngle(pt1,pt2){
+
+   var  x1 = pt1["geometry"]["coordinates"][0];
+   var  x2 = pt2["geometry"]["coordinates"][0];
+   var y1 =  pt1["geometry"]["coordinates"][1];
+   var y2 =  pt2["geometry"]["coordinates"][1];
+   var angle = Math.atan2(y2-y1,x2-x1)*180/Math.PI;
+   ratio = (y2-y1)/(x2-x1);
+   console.log("angle",angle);
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Remove this?

File: twittermap/public/javascripts/cache/services.js [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+                     console.error("Load city data failure");
+                 });
+             });
+
+             return deferred.promise();
+          }
+}
+
+/*Moved Right is true,Moved left is false
+Moved Up is true,Moved Down is false*/
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Please put a comment here. What does this function do?

File: twittermap/public/javascripts/cache/services.js [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+   ratio = (y2-y1)/(x2-x1);
+   console.log("angle",angle);
+   return angle;
+}
+
+
+
+
+
+function prefetch(currentRequestCentroid,curReqBBox){
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Please put a comment here. What does this function do?

File: twittermap/public/javascripts/cache/services.js [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+        //Make 10 slices
+        var ShiftX = distanceX/10;
+        var ShiftY = distanceY/10;
+        var start = 0;
+        var stop= distanceX;
+        var removeItems;
+        var slicedXline,slicedYline,cutPointX,cutPointY,cutBbox,remove_search;
+        var moveX = ShiftX;
+        var moveY = ShiftY;
+        var ti = performance.now();
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Remove this performance measurement related part.

File: twittermap/public/javascripts/cache/services.js [hunk 8]
```diff
@@ -0,0 +1,575 @@
...
+        var count = 0;
+        while(deletedPolygonCount<targetDeleteCount){
+
+
+            slicedXline = turf.lineSliceAlong(Xline, start, moveX, 'miles');
+            slicedYline = turf.lineSliceAlong(Yline, start, moveY, 'miles');
+            cutPointX = slicedXline["geometry"]["coordinates"][1][0];
+            cutPointY = slicedYline["geometry"]["coordinates"][1][1];
+            if(Direction == "NE"){
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Trend = NE, Fathest Region = NE, we can’t delete SW, NE. Pick the farthest one from NW, SE.
Trend = SE, Fathest Region = SE, we can’t delete NW, SE. Pick the farthest one from NE, SW.
Trend = NW, Fathest Region = NW, we can’t delete SE...
Trend = SW, Fathest Region = SW, we can’t delete NE...

File: twittermap/public/javascripts/cache/services.js (hunk 7 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: removed

File: twittermap/public/javascripts/cache/services.js [hunk 9]
```diff
@@ -0,0 +1,575 @@
...
+        var newminY = minY-maxDelta;
+        newbbox = [minX,newminY,newmaxX,maxY];
+        user_direction = "SE";
+
+    }else if( angle>90 && angle<=180) {
+        //NW
+
+        var newminX = minX - maxDelta;
+        var newmaxY = maxY + maxDelta;
+        if(newminX>minX){
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: removed

File: twittermap/public/javascripts/cache/services.js [hunk 10]
```diff
@@ -0,0 +1,575 @@
...
+
+    }
+    else if(angle>-45 && angle<=45){
+        //line2-RIGHT
+         x2 = curReqBBox[2];
+         y2 = ratio*(x2-x1)+y1;
+
+    }
+    else if(angle>=-135 && angle<=-45){
+        //line3-DOWN
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: changed

File: twittermap/public/javascripts/cache/services.js [hunk 11]
```diff
@@ -0,0 +1,575 @@
...
+                 extraBounds = "city/"+newMBR[3]+"/"+newMBR[1]+"/"+newMBR[2]+"/"+newMBR[0];
+                 $http.get(extraBounds).success(function(data) {
+
+                     insertIntoTree(data.features,RequestPolygonWithPrefetch).done(function(){
+
+                         data_response = data;
+
+                         if(cachedRegion == undefined)
+                         {cachedRegion = currentRequestPolygon;
+                             displayCacheRegion = cachedRegion;}
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: removed

File: twittermap/public/javascripts/cache/services.js [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+
+
+   if(typeof cachedRegion != "undefined" && typeof turf.difference(currentRequestPolygon,cachedRegion) == "undefined") {
+          //cache HIT
+
+         var result = cachedCityPolygonTree.search(item);
+         data_response = turf.featureCollection(result);
+         cacheHitCount += 1;
+         Hit = true;
+         RequestPolygonWithPrefetch = currentRequestPolygon;
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: removed

File: twittermap/public/javascripts/cache/services.js [hunk 13]
```diff
@@ -0,0 +1,575 @@
...
+var deletedPolygonCount = 0;
+var preFetchDistance = 25;
+var previousRequestsData = [];
+var movementHistory = [];
+var noOfHistories = 5;
+var previousRequestCentroid;
+var cacheHitCount = 0;
+var cacheMissCount = 0;
+var currentReqMBR;
+var Hit = false;
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: removed not neccessary used in demo only

File: twittermap/public/javascripts/cache/services.js (hunk 4 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: removed

File: twittermap/public/javascripts/cache/services.js (hunk 6 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Comment added

File: twittermap/public/javascripts/cache/services.js (hunk 3 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: added

File: twittermap/public/javascripts/app.js (hunk 1 above)
Comment (vidhya567) [inline twittermap/public/javascripts/app.js]: removed

**Comment by waans11** [comment]
@JavierJia : if you don't have any comments, I think we can merge this now.

**Comment by JavierJia** [comment]
Sure. I will let you decide :-)

**Comment by chenlica** [comment]
Here's a warning: "codacy/pr — Not so good... This pull request quality could be better."

After the merge, let's push it to the TwitterMap and let Jianfeng do the refactoring of the codebase.

**Comment by waans11** [comment]
Then, I think we can merge this patch.

**Pull request merged**